Docs updates (#2081)
vyperlang/vyper: docs/types.rst

.. index:: ! bool, ! true, ! false

Boolean
-------

**Keyword:** ``bool``

A boolean is a type to store a logical/truth value.

Values
******

The only possible values are the constants ``True`` and ``False``.

> Operators

Author: That relates to boolean operators, not booleans.  Not sure where it'd belong, somewhere in the control flow section probably?

``x * y``      Multiplication
``x / y``      Division
``x % y``      Modulo
=============  ==========================================

``x`` and ``y`` must be of the type ``decimal``.

.. _address:

Address
-------

**Keyword:** ``address``

The address type holds an Ethereum address.

Values
******

> An address type can hold an Ethereum address which equates to 20 bytes or 160 bits. Address literals must be written in hexadecimal notation with a leading ``0x``. All addresses must be `checksummed <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md>`_.

Author: Good catch

.. index:: !bytes

Fixed-size Byte Arrays
----------------------

**Keyword:** ``Bytes``

> A byte array with a fixed size.

Reviewer: `Bytes` is a dynamic type

Author: I think the idea here is that because the size in memory is fixed, the type itself is "fixed size".

Reviewer: "Fixed" strikes me as another way to say "static", which they're definitely not static.

Reviewer: How about just "Byte Arrays"?

.. code-block:: python

    bytes_string: Bytes[100] = b"\x01"
    hex_bytes: Bytes[100] = 0x01
    binary_bytes: Bytes[100] = 0b00000001

.. index:: !string

Fixed-size Strings
------------------

**Keyword:** ``String``
Fixed-size strings can hold strings with equal or fewer characters than the maximum length of the string.
On the ABI level the Fixed-size bytes array is annotated as ``string``.

.. code-block:: python

    example_str: String[100] = "Test String"

.. index:: !reference

Reference Types
===============

Reference types do not fit into 32 bytes. Because of this, copying their value is not as feasible as
with value types. Therefore only the location, i.e. the reference, of the data is passed.

.. index:: !arrays

Fixed-size Lists
----------------

Fixed-size lists hold a finite number of elements which belong to a specified type.

Syntax
******

Lists can be declared with ``_name: _ValueType[_Integer]``. Multidimensional lists are also possible.

.. code-block:: python

    # Defining a list
    exampleList: int128[3]

    # Setting values
    exampleList = [10, 11, 12]
    exampleList[2] = 42

    # Returning a value
    return exampleList[0]

.. _types-struct:

> Structs

Reviewer: This section could be more filled out, or link to a better section

-------

Structs are custom defined types that can group several variables.

Syntax
******

Struct members can be accessed via ``struct.argname``.

.. code-block::

    # Defining a struct
    struct MyStruct:
        value1: int128
        value2: decimal

    # Declaring a struct variable
    exampleStruct: MyStruct = MyStruct({value1: 1, value2: 2})

    # Accessing a value
    exampleStruct.value1 = 1

.. index:: !mapping

Mappings
--------

> Mappings are `hash tables <https://en.wikipedia.org/wiki/Hash_table>`_ that are virtually initialized such that every possible key exists and is mapped to a value whose byte-representation is all zeros: a type's :ref:`default value <types-initial>`.

Reviewer: Maybe mention that the default value is the same value produced by `empty(type)`

Author: There is a link to the default values section where `empty` is mentioned.

The key data is not stored in a mapping, instead its ``keccak256`` hash used to look up a value. For this reason mappings do not have a length or a concept of a key or value being "set".

.. note::
    Mappings are only allowed as state variables.

Syntax
******

Mapping types are declared as ``HashMap[_KeyType, _ValueType]``.

> * ``_KeyType`` can be any base or bytes type. Mappings, interfaces or structs are not support as key types.

Reviewer: Interfaces should eventually be allowed

.. _type_conversions:

Type Conversions
================

> All type conversions in Vyper must be made explicitly using the built-in ``convert(a, b)`` function. Currently, the following type conversions are supported:

Reviewer: ```suggestion
All type conversions in Vyper must be made explicitly using the built-in ``convert(a: atype, btype) -> btype`` function. Currently, the following type conversions are supported:
```

================ ================== ============================== =================================

> Initial Type (a) Converted Type (b) Allowable Values               Additional Notes

Reviewer: ```suggestion
In (``atype``) Out (``btype``) Allowable Values               Additional Notes
```

Reviewer: or just 
```suggestion
``atype`` ``btype`` Allowable Values               Additional Notes
```

================ ================== ============================== =================================
``bool``         ``decimal``        All                            ``0.0`` or ``1.0``
``bool``         ``int128``         All                            ``0`` or ``1``
``bool``         ``uint256``        All                            ``0`` or ``1``
``bool``         ``bytes32``        All                            ``0x00`` or ``0x01``
``decimal``      ``bool``           All                            Returns ``a != 0.0``
``decimal``      ``int128``         All                            Value is truncated

> ``decimal``      ``uint256``        ``0`` ... ``MAX_DECIMAL``      Value is truncated

Reviewer: ```suggestion
``decimal``      ``uint256``        ``a >= 0.0``      Value is truncated
```

``decimal``      ``bytes32``        All
``decimal``      ``Bytes``          All
``int128``       ``bool``           All                            Returns ``a != 0``
``int128``       ``decimal``        All

> ``int128``       ``uint256``        ``0`` ... ``MAX_INT128``       Cannot convert negative values

Reviewer: ```suggestion
``int128``       ``uint256``        ``a >= 0``       Cannot convert negative values
```

``int128``       ``bytes32``        All
``int128``       ``Bytes``          All
``uint256``      ``bool``           All                            Returns ``a != 0``

> ``uint256``      ``decimal``        ``0`` ... ``MAX_DECIMAL``

Reviewer: ```suggestion
``uint256``      ``decimal``        ``a <= MAX_DECIMAL``
```

> ``uint256``      ``int128``         ``0`` ... ``MAX_INT128``

Reviewer: ```suggestion
``uint256``      ``int128``         ``a <= MAX_INT128``
```

.. _type_conversions:

> Type Conversions

Reviewer: This table uses older [constant names](https://vyper.readthedocs.io/en/release-0.2.0/constants-and-vars.html#built-in-constants), and also doesn't display all on one page
